### ISSUE TITLE
[discussion] SpectralNorm behavior in eval() and at remove()

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#8208]
1. `net.eval()`:
Currently, even in eval time, [our forward pre hook](https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/spectral_norm.py#L43-L46) still does power iteration and **updates `weight_u`**. This means that the function the network represents is changing as we test on more and more images! Certainly this shouldn't be desired, despite that the [original code](https://github.com/pfnet-research/chainer-gan-lib/blob/master/common/sn/sn_convolution_2d.py#L68-L104) has this behavior. The question is what the correct behavior is. Should we just skip the power iteration, i.e. only compute `_v` and not a new `_u`?

2. `sn_handle.remove()`:
We currently [set the weight parameter to the saved `weight_orig`](https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/spectral_norm.py#L37-L41), i.e., the un-normalized weight, at removing SN. I don't think that this is correct either, because the effective weight is now `weight_orig / sigma(weight)` and this removal drastically changes the network behavior. I propose to set the weight parameter to result of `compute_weight()`, and maybe also with power iteration skipped.

@crcrpar @t-vi @fmassa  Let me know what you guys think. :)

**Comment by crcrpar** [comment]
Hi,

1. I agree. Skipping the computation of new `_u`, i.e. your proposition sounds good.
2. `remove` should remove the forward hook and register the latest weight, i.e. the normalized one.

**Comment by t-vi** [comment]
Thanks for looking into this!
I agree that one should expect the weight to "be fixed" to the computed weight in both cases.
In terms of practically achieving this, my ideas would be:

1. The current behaviour seems suboptimal indeed. I would expect the `.eval`ed network to give the same weight as the previous iteration. The current code does not make this easy to achieve except for caching `v`. Would having an additional half-iteration for `v` at the end of the iteration (in training as well) be a justifiable departure from a "pure" implementation? Mathematically it should be sound.
2.  Once the `eval` mode is solved, `sn_handle.remove()` should probably set the weight to an `eval`-mode computation of the weight, as this seems the generically right thing to do.

Once you decide on a course: Do you want to implement this or should I? Now that my open pull requests have deminished, I'd love to add some fresh ones. :slightly_smiling_face:

**Comment by SsnL** [comment]
@t-vi I'm fine with caching `_v`. Admittedly `_v` can be much larger than `_u` (especially when on conv weights, but I don't think that the memory increase will be too bad. Please go ahead implementing this if you want to. But be aware of #8170 which might have merge conflict with your patch.

**Comment by SsnL** [comment]
@t-vi That PR is merged. Go for the changes if you want to! 

Another thing we should do for SN is #7865 . Currently we assume that weight.dim(1) is the output feature dimension. However, this is not true for the ConvTranspose cases (should be dim(0))! This is crucial as ConvTs are commonly used in GANs. Fortunately with SN we are given the modules and should be able to detect this case. I'd be more than happy to review if you also want to fix this. Ideally we should also have a `dim=` argument. :)

**Comment by t-vi** [comment]
Thanks for the head up. I should have a PR tomorrow in the evening.

**Comment by SsnL** [comment]
@t-vi Thanks! No rush.
